### PR TITLE
TPM: build issue fixing

### PR DESCRIPTION
**Pull request by arm64b**
Fix the build issue on arm64 and amd64.

Signed-off-by: Dennis Chen <dennis.chen@arm.com>

### CODE REVIEW

**Comment by arm64b** [comment]
Hi @dm0- , any comments?

**Comment by dm0-** [comment]
Sorry, I haven't had a chance to test this yet.  What environment are you using to build this where you saw issues?  I don't see any TPM-related warnings during the CoreOS GRUB build.

**Comment by arm64b** [comment]
I build the CoreOS GRUB on both x86_64 and arm64 platform with ubuntu 16.04 installed, I pull the `master` branch at commit `84a4fd2f4c35a45b66a6ca4ef455d999f8b05b55`.
```
#./autogen.sh
#./configure --libdir=/grub-lib --with-platform=efi
#make
```
The error message is:
```
In file included from kern/efi/tpm.c:3:0:
kern/efi/tpm.c: In function ‘grub_tpm1_log_event’:
kern/efi/tpm.c:202:52: error: passing argument 2 of ‘tpm->log_extend_event’ makes integer from pointer without a cast [-Werror=int-conversion]
   status = efi_call_7 (tpm->log_extend_event, tpm, buf, (grub_uint64_t) size,
...
```

**Comment by arm64b** [comment]
Hi @dm0-, any feedback or something I can do here?

**Comment by dm0-** [comment]
Sorry, I haven't had the time to properly go through the TPM GRUB code to verify this, and I will probably be tied up for at least another month (unless someone else wants to review this).

If I may ask, why are you building the CoreOS GRUB repo on Ubuntu?  If it's TPM support that you want, @mjg59 has been sending updated patches to the upstream GRUB developer mailing list that use the verifiers API.  It might be cleaner when handling future changes to base your project on these updated patches.

**Comment by arm64b** [comment]
Actually we use the CoreOS GRUB on the LinuxKit project, so the build error is also in the Alpine distribution. We do want the TPM support and also notice the patches sent by @mjg59 , but that patches in based on the 'verifier' branch of the upstream where there are some build issues after applied mjg59's patch, so I decide to use CoreOS as the solution given those patches haven't been merged into the upstream.

File: grub-core/kern/efi/tpm.c [hunk 1]
```diff
@@ -175,7 +175,7 @@ grub_tpm1_log_event(grub_efi_handle_t tpm_handle, unsigned char *buf,
 		    grub_size_t size, grub_uint8_t pcr,
 		    const char *description)
 {
-  Event *event;
+  TCG_PCR_EVENT *event;
```
Comment (dm0-) [inline grub-core/kern/efi/tpm.c]: What does this change affect?  Both types appear to be equivalent.  If you want to drop `Event` usage, the unused structure definition might as well be dropped, too.  (And maybe have that in a separate commit.)

File: grub-core/kern/efi/tpm.c [hunk 2]
```diff
@@ -188,15 +188,15 @@ grub_tpm1_log_event(grub_efi_handle_t tpm_handle, unsigned char *buf,
...
-  event->eventtype = EV_IPL;
-  event->eventsize = grub_strlen(description) + 1;
-  grub_memcpy(event->event, description, event->eventsize);
+  event->PCRIndex = pcr;
+  event->EventType = EV_IPL;
+  event->EventSize = grub_strlen(description) + 1;
+  grub_memcpy(event->Event, description, event->EventSize);
 
   algorithm = TCG_ALG_SHA;
   status = efi_call_7 (tpm->log_extend_event, tpm, buf, (grub_uint64_t) size,
```
Comment (dm0-) [inline grub-core/kern/efi/tpm.c]: I've tried building GRUB with `-Werror=int-conversion`, but my GCC doesn't have a problem with this.

If it fixes your build, I think that it would be more correct to explicitly cast `buf` to `grub_efi_physical_address_t` here than to change the protocol structures.  Also, you might have to do the same below for `hash_log_extend_event`.

File: grub-core/kern/efi/tpm.c (hunk 1 above)
Comment (arm64b) [reply grub-core/kern/efi/tpm.c]: Ah, I will try to drop the unused Event data structure in a separate PR. The data type of all the members of the Event and TCG_PCR_EVENT is the same, but the name is different. The following code just use the members from TCG_PCR_EVENT Type.

File: grub-core/kern/efi/tpm.c (hunk 2 above)
Comment (arm64b) [reply grub-core/kern/efi/tpm.c]: yes, of course. it should be `(grub_efi_physical_address_t)buf`, the same to `hash_log_extend_event`

File: include/grub/efi/tpm.h [hunk 3]
```diff
@@ -69,7 +69,7 @@ struct grub_efi_tpm_protocol
 					    grub_efi_uint32_t TpmOutputParameterBlockSize,
 					    grub_efi_uint8_t *TpmOutputParameterBlock);
   grub_efi_status_t (*log_extend_event) (struct grub_efi_tpm_protocol *this,
-					 grub_efi_physical_address_t HashData,
+					 grub_efi_uint8_t *HashData,
```
[the remark this answers is not in the document]
Comment (arm64b) [reply include/grub/efi/tpm.h]: Will reserve the original prototype and explicitly cast the `buf` to `grub_efi_physical_address_t` as mentioned above.

File: include/grub/efi/tpm.h [hunk 4]
```diff
@@ -129,7 +129,7 @@ struct grub_efi_tpm2_protocol
 				      grub_efi_boolean_t *EventLogTruncated);
   grub_efi_status_t (*hash_log_extend_event) (struct grub_efi_tpm2_protocol *this,
 					      grub_efi_uint64_t Flags,
-					      grub_efi_physical_address_t *DataToHash,
+					      grub_efi_uint8_t *DataToHash,
```
[the remark this answers is not in the document]
Comment (arm64b) [reply include/grub/efi/tpm.h]: So we will have another commit to correct the `DataToHash` type as `grub_efi_physical_address_t DataToHash`

**Comment by arm64b** [comment]
@dm0-  According to your comments, I re-push 3 separate commits here.

File: grub-core/kern/efi/tpm.c [hunk 5]
```diff
@@ -161,21 +161,12 @@ grub_tpm_execute(PassThroughToTPM_InputParamBlock *inbuf,
   }
 }
 
-typedef struct {
-	grub_uint32_t pcrindex;
-	grub_uint32_t eventtype;
-	grub_uint8_t digest[20];
-	grub_uint32_t eventsize;
-	grub_uint8_t event[1];
-} Event;
```
Comment (dm0-) [inline grub-core/kern/efi/tpm.c]: I think this deletion should be in the same commit with changing `event` to `TCG_PCR_EVENT` below since they seem to be parts of the same change (dropping a redundant struct).  That can be in a separate commit from the `buf` casting change.  This is minor, though, and I'm okay with merging this without it.

**Comment by lucab** [comment]
I had a look at the type-casting and this patch seems ok, however I don't have any specific prior knowledge of grub-tpm.
